### PR TITLE
Haskell checks on ci

### PR DESCRIPTION
**Pull request by sashasashasasha151**
## Related issue(s)

https://issues.serokell.io/issue/EDNA-11

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)

#### Stylistic guide (mandatory)

- [ ] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [ ] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: backend/default.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
+{ haskell-nix, runCommand, lib }:
+let
+  project = haskell-nix.stackProject {
+    src = haskell-nix.haskellLib.cleanGit {
+      src = haskell-nix.cleanSourceHaskell { src = ./.; };
+    };
+    modules = [
+      {
+        doHaddock = false;
```
Comment (gromakovsky) [inline backend/default.nix]: I think we want to do Haddock btw.

File: flake.lock [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+        "rev": "91e934fd9e1e06044f75a3e3ad998813c2a90fb5",
+        "type": "github"
+      },
+      "original": {
+        "owner": "serokell",
+        "repo": "deploy-rs",
+        "type": "github"
+      }
+    },
+    "deploy-rs": {
```
Comment (gromakovsky) [inline flake.lock]: I know it's not ready for review yet, but keep in mind that many things here are unused (AFAIU) and should be removed, e. g. I doubt we use `deploy-rs`.

File: .buildkite/pipeline.yml [hunk 3]
```diff
@@ -1,4 +1,7 @@
 steps:
-  - label: check trailing whitespace
+#  - label: check trailing whitespace
+#    commands:
+#      - .buildkite/check-trailing-whitespace.sh
```
Comment (gromakovsky) [inline .buildkite/pipeline.yml]: Did it stop working?

**Comment by gromakovsky** [comment]
We've decided not to do the check for cabal files in this PR (unless we find a solution quickly).

File: .buildkite/pipeline.yml [hunk 4]
```diff
@@ -1,4 +1,16 @@
 steps:
   - label: check trailing whitespace
     commands:
-    - .buildkite/check-trailing-whitespace.sh
+      - .buildkite/check-trailing-whitespace.sh
+  - label: backend build
```
Comment (gromakovsky) [inline .buildkite/pipeline.yml]: What is `backend build` exactly? We have `backend lib`, `backend test` and `backend build`. Is `backend build` supposed to build the library, test, everything or something else? I think there is some redundancy here.

File: backend/.#default.nix [hunk 5]
```diff
@@ -0,0 +1 @@
+balsoft@AMD-Workstation.15930:1613204272
```
Comment (gromakovsky) [inline backend/.#default.nix]: Don't forget to remove it.

File: backend/default.nix [hunk 6]
```diff
@@ -0,0 +1,44 @@
+{ haskell-nix, runCommand, lib, hlint }:
+let
+  ignoreFilter = path: type:
```
Comment (gromakovsky) [inline backend/default.nix]: Do you understand why we need this `ignoreFilter`?

File: backend/default.nix [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    in
+      !(
+        baseName == ".gitignore" ||
+        hasPrefix "resources" relPath ||
+        hasSuffix ".md" baseName ||
+        hasSuffix ".nix" baseName
+      );
+
+  # Set a constant name for the src closure
+  source = let
```
Comment (gromakovsky) [inline backend/default.nix]: AFAIU this `source` thing and the above `ignoreFilter` are only used to run `hlint`, I see it used only in `hlint-html` (and btw, does `${source}` there refer to this `source` or something else?). Is it indeed needed only for `hlint` or is it implicitly used somewhere else? I just Ctrl-F'ed `source`.

I have an alternative solution: just call `hlint` from `pipeline.yaml` using `nix run`. Pros:
1. No need to have all this extra code in this file (`source`, `ignoreFilter`). Though maybe I am wrong and it's needed for something else as well.
2. Less steps involved, easier to understand what's going on, easier to find what arguments are passed and where to change them.

AFAIU there is one problem: we don't know how to make `nix run` work with flakes. If that's the case and we don't figure it out quickly, I am okay with changing it later.

File: backend/default.nix [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+          doHaddock = true;
+          package.ghcOptions = "-Werror";
+        };
+      }
+    ];
+  };
+  library = project.edna.components.library;
+  server = project.edna.components.exes.edna-server;
+  test = project.edna.checks.edna-test;
+  hlint-html = runCommand "hlint.html" {} "${hlint}/bin/hlint ${source} --no-exit-code --report=$out -j";
```
Comment (gromakovsky) [inline backend/default.nix]: 1. What is `hlint.html`? Do you understand?
2. Why do we pass `--no-exit-code`? This way it seems very unlikely to notice when it reports anything. If we don't want CI to fail when `hlint` fails, we can mark the corresponding step as optional in the CI config. Just set `soft_fail: true`. I think it's better because this way it will be easier to notice that `hlint` reported something.

**Review by gromakovsky — CHANGES_REQUESTED** [review]

File: pipeline.nix [hunk 9]
```diff
@@ -0,0 +1 @@
+(import ./.).pipelineFile
```
Comment (gromakovsky) [inline pipeline.nix]: Is this file needed?

File: flake.nix [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+      ciSystems = [ "x86_64-linux" ];
+
+      checks = builtins.mapAttrs (system: _:
+        with packagesFor system; {
+          backend-test = backend.test;
+          backend-hlint = backend.hlint;
+        }) nixpkgs.legacyPackages;
+
+      pipelineFile = common-infra.mkPipelineFile self;
```
Comment (gromakovsky) [inline flake.nix]: And do we need this?

File: flake.nix [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+      devShell = builtins.mapAttrs (system: packages:
+        let pkgs = nixpkgs.legacyPackages.${system};
+        in pkgs.mkShell {
+          inputsFrom = [ packages.backend-lib ];
+          buildInputs = with pkgs.haskellPackages; [ cabal-install hpack ];
```
Comment (gromakovsky) [inline flake.nix]: I don't use dev shell and I am not sure anybody needs it, but if we decide to keep it, probably we don't need `hpack` here because we don't use it (no `package.yaml`).

File: flake.nix [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+      url = "github:input-output-hk/stackage.nix";
+      flake = false;
+    };
+    common-infra.url = "github:serokell/common-infra";
+  };
+
+  outputs = { self, nixpkgs, haskell-nix, hackage, stackage, common-infra }:
+    let
+      packagesFor = system: rec {
+        pkgs = nixpkgs.legacyPackages.${system}.extend
```
Comment (gromakovsky) [inline flake.nix]: What is the semantic of this `pkgs`?

File: flake.nix [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+            haskell-nix.overlay
+          ]);
+        backend = pkgs.callPackage ./backend { };
+      };
+    in {
+      packages = builtins.mapAttrs (system: _:
+        with packagesFor system; {
+          backend-lib = backend.library;
+          backend = backend.server // {
+            meta.modulePath = [ "services" "edna" "backend" ];
```
Comment (gromakovsky) [inline flake.nix]: What is `meta.modulePath`?

File: flake.nix [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+          };
+        }) nixpkgs.legacyPackages;
+
+      ciSystems = [ "x86_64-linux" ];
```
Comment (gromakovsky) [inline flake.nix]: Btw, it seems very unlikely that we will have more than one `ciSystem` in any foreseeable future, so this file can be simplified, we don't need to iterate over systems and pass them to functions, we can have a single constant `ciSystem`.

File: flake.lock [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+        "rev": "ba87ffb84bc4158ada14908bddee9df2c81ac31b",
+        "type": "github"
+      },
+      "original": {
+        "owner": "input-output-hk",
+        "repo": "haskell.nix",
+        "type": "github"
+      }
+    },
+    "naersk": {
```
Comment (gromakovsky) [inline flake.lock]: https://github.com/nmattia/naersk says "Nix support for building cargo crates". Do we need it?

**Comment by gromakovsky** [comment]
@mkaito I have a question to you.
In many repos we often do things like `nix run -f ci.nix pkgs.reuse -c reuse lint`. By passing `pkgs.reuse` here we ensure that `reuse` is available in `$PATH`. To make it work we need to define `pkgs` in `ci.nix`, so that it has `reuse` attribute. In `ci.nix` we have something like this:
```
  sources = import ./nix/sources.nix;
  haskellNix = import sources."haskell.nix" {
    sourcesOverride = { hackage = sources."hackage.nix"; stackage = sources."stackage.nix"; };
  };
  pkgs = import sources.nixpkgs haskellNix.nixpkgsArgs;
```

So AFAIU `pkgs` is `nixpkgs` and its revision is taken from `nix/sources.json`. That works well when we use `niv`.
I like this `nix run -f ci.nix INSTALLABLE -c bla-bla-bla` feature and I think we want to use it in this repo. But it's not clear what's the best way to do the same with flakes instead of `niv`.
First of all, since we have `default.nix` instead of `ci.nix`, we don't need to pass `-f ci.nix`, that's clear to me.
But it's not clear how to access nixpkgs like we have `pkgs` in the approach with `niv`. We have `defaultNix` in `default.nix`, so AFAIU its contents is fully defined by `flake.nix`. Since we have `nixpkgs` in flake's `outputs`, can we just write `nix run nixpkgs.reuse`? Or should we do it differently?

P. S. Example with `niv`: https://github.com/tqtezos/baseDAO

**Comment by gromakovsky** [comment]
> Since we have `nixpkgs` in flake's `outputs`

Ah, oh, sorry, it's not in `outputs`, it's in **arguments** of `outputs`, so `outputs` expects `nixpkgs` to be passed to it.

I've just tested my hypotheses. First I tried to just do `nix run nixpkgs.hlint -c hlint`, see [here](https://buildkite.com/serokell/edna/builds/86#fe1be568-d905-49d6-bee1-97dcbb59bad9). It said:
> attribute 'nixpkgs' in selection path 'nixpkgs.hlint' not found

Then I tried to also add `nixpkgs = nixpkgs;` (not sure if it makes sense), but got the same message [here](https://buildkite.com/serokell/edna/builds/87#ddec1278-729e-4b0f-b568-60fb4b577aed). You can also see commits I made, they are available from buildkite.

I don't want to go further with experiments, will wait for Chris to respond.

File: flake.nix (hunk 11 above)
Comment (mkaito) [reply flake.nix]: `devShell` is the Flake equivalent of `shell.nix` -- and yes, I use it all the time.

File: flake.lock (hunk 15 above)
Comment (mkaito) [reply flake.lock]: It's used by `deploy-rs`. `flake.lock` contains all transitive dependencies for all inputs in the flake. `deploy-rs` is a transitive dependency of `common-infra` which is an input to this flake.

Comment (gromakovsky) [reply flake.lock]: AFAIU we don't use `deploy-rs` either, but I see that it's a dependency of `common-infra`, so ok. It's a bit poor granularity (that we need to mention so many packages just for a small piece of `common-infra`), but I guess we will end up using `deploy-rs` anyway, so it's not a problem.

**Comment by mkaito** [comment]
The command `nix run` changed semantics in 3.0. In 2.4 it was a replacement for `nix-shell -p blah --run whatever`, but in 3.0 it does something else entirely.

It relies on a flake concept called "apps", where you define executables in a flake output that can be run with `nix run` without installing. For example, you can `nix run github:serokell/deploy-rs` to run that program directly from git master without installing it.

What you're looking for is now called `nix shell`.

This will do what you want.

```sh
$ nix shell nixpkgs#hlint -c hlint
```

Note the `nixpkgs#hlint` reference. This puts the package `hlint` from the flake `nixpkgs` in `$PATH` and then calls command `hlint`. The `nixpkgs` flake is defined in the flake registry as `github:nixos/nixpkgs`.

Generally speaking, all our inputs should be declared in the flake, and everything should be run through the flake. The `mkPipelineFile` function will build a sepate CI step for each package, check, and deployment attribute in the flake, so you should just plug things into `checks` as necessary. I'm not sure it supports custom extra steps, but that can probably be added if necessary.

The files `default.nix` and `shell.nix` contain compatibility shims to allow Nix 2.4 to read flake outputs, so you can `nix-build` and `nix-shell` without installing 3.0.

**Comment by gromakovsky** [comment]
> The `mkPipelineFile` function will build a separate CI step for each package

What do you mean by "build a separate CI step"? Does it generate `.pipeline.yml`? Is it supposed to be appended to the main `.pipeline.yml`? Or replace it? Or what? How does it generate steps, how does it know what options to specify for steps, like `soft_fail`, `if`, etc.?

**Comment by gromakovsky** [comment]
> This will do what you want.

Ok, let me test it in my branch.

**Comment by mkaito** [comment]
> Does it generate .pipeline.yml?

Yes.

> How does it generate steps

By reading the `packages` and `checks` outputs in the flake and generating one CI step for each. Here's what it currently generates (`nix build .#pipelineFile && cat result | jq`):

```json
{
  "steps": [
    {
      "agents": [],
      "command": "nix-build -A packages.x86_64-linux.backend-lib",
      "label": "Build backend-lib"
    },
    {
      "agents": [],
      "command": "nix-build -A packages.x86_64-linux.backend-server",
      "label": "Build backend-server"
    },
    {
      "agents": [],
      "command": "nix-build --no-out-link -A checks.x86_64-linux.backend-hlint",
      "label": "backend-hlint"
    },
    {
      "agents": [],
      "command": "nix-build --no-out-link -A checks.x86_64-linux.backend-test",
      "label": "backend-test"
    },
    "wait"
  ]
}


```

> how does it know what options to specify for steps

It doesn't. It just runs what you tell it to run. For example, `checks.backend-hlint` is defined in `backend/default.nix` as such:

```nix
{
  hlint-html = runCommand "hlint.html" {} "${hlint}/bin/hlint ${source} --no-exit-code --report=$out -j";
}
```

File: flake.lock (hunk 15 above)
Comment (mkaito) [reply flake.lock]: `common-infra` has 2 nix functions: `mkNode` to define deployment nodes, and `mkPipelineFile` that generates CI steps which uses the former to generate deployment steps. So we're technically using half of `common-infra` :stuck_out_tongue: 

Packages mentioned in `flake.lock` don't necessarily end up in any of the closures, and since Nix is lazy, they don't even get evaluated unless you actually use them somewhere. The lockfile is eager and will contain all transitive locks regardless of whether they are used. It's just extra noise, that's all.

File: backend/default.nix (hunk 8 above)
Comment (mkaito) [reply backend/default.nix]: 1. It's the name of the output file generated in the closure for the `hlint-html` derivation. It's an arbitrary name. The important part is the closure itself and its hash, which depends on the inputs defined in `source`.

2. No idea. This looks like what Yegor wrote years ago for Disciplina. I suspect it has just been cargo culted ever since. Feel free to adjust the invocation of `hlint` as you see fit. Just make sure you put something in `$out` somehow.

File: .buildkite/pipeline.yml (hunk 4 above)
Comment (mkaito) [reply .buildkite/pipeline.yml]: It builds `backend.server`, according to the way things are declared now. This file won't be used either way, as we'll use the generate pipeline steps.

File: backend/default.nix (hunk 6 above)
Comment (mkaito) [reply backend/default.nix]: Cargo culted from something I wrote in 2017. It's not necessary, and I'm removing it.

File: flake.nix (hunk 10 above)
Comment (mkaito) [reply flake.nix]: We do, actually.

File: flake.nix (hunk 12 above)
Comment (mkaito) [reply flake.nix]: Flakes define derivations for all known systems. In the context of a flake, `nixpkgs` isn't a set of packages. It's a set of sets of packages. So you need to use the set for the current (for some definition of "current") system.

File: flake.nix (hunk 13 above)
Comment (mkaito) [reply flake.nix]: No idea. Removed it.

File: flake.nix (hunk 14 above)
Comment (mkaito) [reply flake.nix]: This isn't used anywhere. `mkPipelineFile` only generates steps for `x86_64-linux`.

**Comment by gromakovsky** [comment]
So does it replace `.pipeline.yml`? Can we refrain from doing that please and just write `.pipeline.yml` manually? I don't like it for many reasons:
1. If I just want to do something simple on CI (e. g. call some `.sh` script), I need to go to some nix file and do it there. That's overhead and entry barrier.
2. Developers generally don't like dealing with nix, so each use of nix should be justified. Here we just move some configuration from yaml to nix files, so I don't see a big benefit.
3. Our CI approach changes every year, some details change every month. This `mkPipelineFile` thing is new (AFAIU) and not battle-tested (in many repos for a long time). I don't want to use such things in this project where we just need to quickly do what we know how to do and don't have much space for experiments. I am tired in this never-ending changes in the way we handle CI.
4. If we want to specify custom `agents` or `if` or set `soft_fail` or any other option, is there any way to do it? AFAIU you are saying there is none.

> For example, `checks.backend-hlint` is defined in `backend/default.nix` as such:

Yes, and I don't like that, don't understand why it's done this way and proposed to change that. I like
```
  - label: hlint
    commands:
      - nix shell nixpkgs#hlint -c hlint
```

much more.

**Comment by mkaito** [comment]
I've cleaned up some things and removed pipeline generation.

**Comment by gromakovsky** [comment]
@mkaito thank you, that's exactly what I wanted to see. LGTM now. One last question I wanted to ask: I see the things in flake's output are called `packages` and `checks`. Is it flake's conversion or our own? I was thinking that these names are treated specially by `mkPipelineFile`: it passes `--no-out-link ` for `checks` and doesn't do it for `packages`. Is there anything else where these names matter?

If I rename `packages` to `foo` and `checks` to `bar` and update `.pipeline.yml` accordingly, will it work the same way? I am ok with current names and don't suggest changing them, just want to know whether there are any other conventions.

**Review by gromakovsky — APPROVED** [review]
I'm ok with merging it given that:
1. You fix CI so that it passes (obviously).
2. Commits are cleaned up.
3. @sashasashasasha151 reviews latest changes and confirms he's satisfied with them.

**Comment by mkaito** [comment]
> I see the things in flake's output are called packages and checks. Is it flake's conversion or our own?

There are outputs that are treated specially. `packages` are build outputs, while `nix flake check` (need nix 3.0) will eval everything referenced in the flake, build all packages and also all checks. There's other meaningful outputs, such as `nixosModules` and `overlay`. Flakes present a standardized interface to modular repositories of NixOS components. See doc links further down.

> it passes --no-out-link for checks and doesn't do it for packages.

The `result` symlink act as a GC root. As long as something in the store is referenced by a gc root, it will not get collected. We pin closures for packages but not for tests. The rationale being that packages are used for deployment and we don't want to rebuild them if a GC happens at the wrong time.

> Is there anything else where these names matter?

Yeah, although flakes are sparsely documented. Half of what we know we learned by reading the C++ source code of Nix itself. The best documentation I've found is oddly enough `nix flake --help`, for which you need a very recent version of the Nix client installed.

The de-facto standard documentation is the old and closed [RFC 49](https://github.com/NixOS/rfcs/pull/49). There's a half decent [wiki page](https://nixos.wiki/wiki/Flakes) too, and some [tutorial style blog posts](https://www.tweag.io/blog/2020-05-25-flakes/).

The Nix folks keep pussyfooting about Flakes. Meanwhile, the rest of the Nix world has been using them a long time and are quite happy. They actually just moved the target for flake release from 3.0 to 3.5 for no apparent reason other than more bikeshedding. Never change, Nix. Never change...

> If I rename packages to foo and checks to bar and update .pipeline.yml accordingly, will it work the same way?

Yes they will, because we don't use any flake specific features in the current pipeline. The shim in `default.nix` presents things in the "old" way, and we just build the derivations with the right names and that's it. This is necessary because our servers run stable nix. We do try however to make it as easy as possible to switch to flake-native Nix in the future, hopefully without having to rewrite every single pipeline. All of these compat shims are used now because flakes require unstable nix, and we don't want to run unstable nix daemon on our servers.

It becomes important once you use a Nix 3.0 client and want to use things like `nix flake check`, which would eval everything, build all packages (for all architectures!) and also all checks.

Some of the flakes I've written for various projects keep a pinned master build of Nix in the devShell so that we can run things like flake checks directly without a shim, but we haven't quite standardized on how exactly we want to do these things.
